### PR TITLE
feat: support of duplicated data feeds in evm-connector

### PR DESCRIPTION
**Pull request by hatskier**

### CODE REVIEW

File: packages/evm-connector/contracts/core/RedstoneConsumerNumericBase.sol [hunk 1]
```diff
@@ -54,4 +54,59 @@ abstract contract RedstoneConsumerNumericBase is RedstoneConsumerBase {
...
+          break;
+        }
+      }
+
+      // Adding if not included
+      if (!alreadyIncluded) {
+        dataFeedIdsWithoutDuplicates[uniqueDataFeedIdsCount] = dataFeedIdsWithDuplicates[indexWithDup];
+        uniqueDataFeedIdsCount++;
+      }
+    }
```
Comment (cehali) [inline packages/evm-connector/contracts/core/RedstoneConsumerNumericBase.sol:88]: The idea for optimization. Introduce 
```
   Struct dataFeedsWithoutDuplcates {
       string[] ids
       mapping(string => bool) isAlreadyIn
   }
```

Then we can iterate through `dataFeedsWithDuplicates` to check if already in and add it if no or skip if is already in.

File: packages/evm-connector/contracts/core/RedstoneConsumerNumericBase.sol [hunk 2]
```diff
@@ -54,4 +54,59 @@ abstract contract RedstoneConsumerNumericBase is RedstoneConsumerBase {
...
+    // Preparing result values array
+    uint256[] memory valuesWithDuplicates = new uint256[](dataFeedIdsWithDuplicates.length);
+    for (uint256 indexWithDup = 0; indexWithDup < dataFeedIdsWithDuplicates.length; indexWithDup++) {
+      for (uint256 indexWithoutDup = 0; indexWithoutDup < dataFeedIdsWithoutDuplicates.length; indexWithoutDup++) {
+        if (dataFeedIdsWithDuplicates[indexWithDup] == dataFeedIdsWithoutDuplicates[indexWithoutDup]) {
+          valuesWithDuplicates[indexWithDup] = valuesWithoutDuplicates[indexWithoutDup];
+          break;
+        }
+      }
+    }
```
Comment (cehali) [inline packages/evm-connector/contracts/core/RedstoneConsumerNumericBase.sol:108]: Maybe build `mapping(ids => value)` from `valuesWithoutDuplicates` and `dataFeedIdsWithoutDuplicates` and then iterate through `dataFeedIdsWithDuplicates` once and build `valuesWithDuplicates` based on this mapping?
In this scenario, we iterate once through both arrays but we don't have inside loop.

File: packages/evm-connector/contracts/core/RedstoneConsumerNumericBase.sol (hunk 1 above)
Comment (jakub-wojciechowski) [reply packages/evm-connector/contracts/core/RedstoneConsumerNumericBase.sol:88]: There are no mappings in memory, unfortunately :)

File: packages/evm-connector/contracts/core/RedstoneConsumerNumericBase.sol [hunk 3]
```diff
@@ -54,4 +54,59 @@ abstract contract RedstoneConsumerNumericBase is RedstoneConsumerBase {
...
+  function getOracleNumericValuesWithDuplicatesFromTxMsg(bytes32[] memory dataFeedIdsWithDuplicates) internal view returns (uint256[] memory) {
+    // Building an array without duplicates
+    bytes32[] memory dataFeedIdsWithoutDuplicates = new bytes32[](dataFeedIdsWithDuplicates.length);
+    bool alreadyIncluded;
+    uint256 uniqueDataFeedIdsCount = 0;
+
+    for (uint256 indexWithDup = 0; indexWithDup < dataFeedIdsWithDuplicates.length; indexWithDup++) {
+      // Checking if current element is already included in `dataFeedIdsWithoutDuplicates`
+      alreadyIncluded = false;
+      for (uint256 indexWithoutDup = 0; indexWithoutDup < dataFeedIdsWithoutDuplicates.length; indexWithoutDup++) {
```
Comment (jakub-wojciechowski) [inline packages/evm-connector/contracts/core/RedstoneConsumerNumericBase.sol:76]: We can end iteration at uniqueDataFeedIdsCount. It saves ~3k gas in the test example:
for (uint256 indexWithoutDup = 0; indexWithoutDup < uniqueDataFeedIdsCount; indexWithoutDup++)

File: packages/evm-connector/contracts/core/RedstoneConsumerNumericBase.sol [hunk 4]
```diff
@@ -54,4 +54,59 @@ abstract contract RedstoneConsumerNumericBase is RedstoneConsumerBase {
...
+  /**
+   * @dev This function works similarly to the `getOracleNumericValuesFromTxMsg` with the
+   * only difference that it allows to request oracle data for an array of data feeds
+   * that may contain duplicates
+   * 
+   * @param dataFeedIdsWithDuplicates An array of data feed identifiers (duplicates are allowed)
+   * @return An array of the extracted and verified oracle values in the same order
+   * as they are requested in the dataFeedIdsWithDuplicates array
+   */
+  function getOracleNumericValuesWithDuplicatesFromTxMsg(bytes32[] memory dataFeedIdsWithDuplicates) internal view returns (uint256[] memory) {
```
Comment (jakub-wojciechowski) [inline packages/evm-connector/contracts/core/RedstoneConsumerNumericBase.sol:67]: We could probably base it on insertion sort to optimise complexity, but...
It's some extra work and deacrease of readibility, 
Let's leave it for now as it is. 
It's super clear and simple.

File: packages/evm-connector/test/mock-wrapper/duplicated-feeds-lib.test.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+  this.beforeEach(async () => {
+    const ContractFactory = await ethers.getContractFactory(
+      "SampleDuplicatedDataFeeds"
+    );
+    contract = await ContractFactory.deploy();
+    await contract.deployed();
+  });
+
+  it("Should pass", async () => {
```
Comment (jakub-wojciechowski) [inline packages/evm-connector/test/mock-wrapper/duplicated-feeds-lib.test.ts:19]: It'd be nice to ass some more test cases, we can abstract away the const symbolsWithDuplicates array and run on few edge cases (one symbol, empty array, one symbols at the beginning - other at the end, etc]

File: packages/evm-connector/contracts/core/RedstoneDefaultsLib.sol [hunk 6]
```diff
@@ -23,13 +23,15 @@ library RedstoneDefaultsLib {
...
     uint256 receivedTimestampSeconds = receivedTimestampMilliseconds / 1000;
-    bool isFromFuture = block.timestamp < receivedTimestampSeconds;
 
-    if ((block.timestamp + DEFAULT_MAX_DATA_TIMESTAMP_AHEAD_SECONDS) < receivedTimestampSeconds) {
-      revert TimestampFromTooLongFuture(receivedTimestampSeconds, block.timestamp);
-    }
-    if (!isFromFuture && (block.timestamp - receivedTimestampSeconds) > DEFAULT_MAX_DATA_TIMESTAMP_DELAY_SECONDS) {
-      revert TimestampIsTooOld(receivedTimestampSeconds, block.timestamp);
+    if (block.timestamp < receivedTimestampSeconds) {
+      if ((block.timestamp + DEFAULT_MAX_DATA_TIMESTAMP_AHEAD_SECONDS) < receivedTimestampSeconds) {
```
Comment (jakub-wojciechowski) [inline packages/evm-connector/contracts/core/RedstoneDefaultsLib.sol:28]: I like the improvements. Imho it would looked even clearer if we change it to:
(receivedTimestampSeconds - block.timestamp) > DEFAULT_MAX_DATA_TIMESTAMP_AHEAD_SECONDS) 

The conditions and parameters will look symmetric and easy to understand

File: packages/evm-connector/test/libs/redstone-defaults-lib.test.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    ).to.be.revertedWith("TimestampIsTooOld");
+  });
+
+  it("Should revert for timestamp from too long future", async () => {
+    await expect(
+      contract.validateTimestamp(Date.now() + 2 * MILLISECONDS_IN_MINUTE)
+    ).to.be.revertedWith("TimestampFromTooLongFuture");
+  });
+
+  it("Should properly aggregate an array with 1 value", async () => {
```
Comment (jakub-wojciechowski) [inline packages/evm-connector/test/libs/redstone-defaults-lib.test.ts:36]: Let's add some more challenging test cases, like 2 or 4 items in array or an array with a very large outlier or with empty/0 values

**Review by jakub-wojciechowski — APPROVED** [review]

**Review by jakub-wojciechowski — APPROVED** [review]
